### ISSUE TITLE
tempfile implementation on windows doesn't always seem to work

### ISSUE THREAD

**Issue by codylane** [tmatilai/vagrant-proxyconf#220]
Problem Statement:
===============

* [GH-ISSUE-218](https://github.com/tmatilai/vagrant-proxyconf/issues/218#issuecomment-639068961)

Technical Review Synopsis:
====================

* [GH-ISSUE-218](https://github.com/tmatilai/vagrant-proxyconf/issues/218#issuecomment-639142802)

**Comment by chucknelson** [comment]
@codylane - Picking up conversation here - additional info:
* Same error happens with Vagrant 2.2.8 and 2.2.9.
* Using Windows 7 (😭 ) 
* Since it does seem related to a core Ruby feature ([Tempfile](https://docs.ruby-lang.org/en/2.6.0/Tempfile.html)), I wonder if it's related to the embedded Ruby version Vagrant uses during commands - for 2.2.8 they seem to have moved from Ruby 2.4.9 to Ruby 2.6.6.

**Comment by codylane** [comment]
Bummer.  Ok, let's keep digging into this as it may be a ruby thing.

I know how the tempfile support is supposed to work in C but I don't really understand the ruby version all that well.  Let me dig a little in the ruby code base and see what I can grok/understand. 

I'm not much of a windows guy but I'll try and find a windows 7 box and see if I can reproduce this issue.  I'm starting to think this may affect every windows user.

Thanks for sharing the link to the ruby docs as well.  This [line](https://docs.ruby-lang.org/en/2.6.0/Tempfile.html#class-Tempfile-label-Good+practices) is of interest to me


Good practices
===========
Explicit close

When a Tempfile object is garbage collected, or when the Ruby interpreter exits, its associated temporary file is automatically deleted. This means that's it's unnecessary to explicitly delete a Tempfile after use, though it's good practice to do so: not explicitly deleting unused Tempfiles can potentially leave behind large amounts of tempfiles on the filesystem until they're garbage collected. The existence of these temp files can make it harder to determine a new Tempfile filename.

Therefore, one should always call unlink or close in an ensure block, like this:

```
file = Tempfile.new('foo')
begin
   # ...do something with file...
ensure
   file.close
   file.unlink   # deletes the temp file
end
```

#### Unlink after creation

On POSIX systems, it's possible to unlink a file right after creating it, and before closing it. This removes the filesystem entry without closing the file handle, so it ensures that only the processes that already had the file handle open can access the file's contents. It's strongly recommended that you do this if you do not want any other processes to be able to read from or write to the Tempfile, and you do not need to know the Tempfile's filename either.

For example, a practical use case for unlink-after-creation would be this: you need a large byte buffer that's too large to comfortably fit in RAM, e.g. when you're writing a web server and you want to buffer the client's file upload data.

Please refer to unlink for more information and a code example.

## Unlink-before-close

On POSIX systems it's possible to unlink a file before closing it. This practice is explained in detail in the Tempfile overview (section “Unlink after creation”); please refer there for more information.

However, unlink-before-close may not be supported on non-POSIX operating systems. Microsoft Windows is the most notable case: unlinking a non-closed file will result in an error, which this method will silently ignore. If you want to practice unlink-before-close whenever possible, then you should write code like this:

```
file = Tempfile.new('foo')
file.unlink   # On Windows this silently fails.
begin
   # ... do something with file ...
ensure
   file.close!   # Closes the file handle. If the file wasn't unlinked
                 # because #unlink failed, then this method will attempt
                 # to do so again.
end
```

# Vagrant-proxyconf's implementation

```
require 'tempfile'

def tempfile(content)
  Tempfile.new("vagrant").tap do |temp|
    temp.binmode
    temp.write(content)
    temp.close
  end
end
```

I suspect the close is what's causing the problems for windows users.  I have a feeling this may not be an atomic operation on windows.  I'd love to setup a simulation that runs a script 1000 times and reports the number of times of pass and fail of each run. Let me see if I can whip something up.

**Comment by codylane** [comment]
## Simulation Code

I've not had much luck setting up Windows and I've lost a few hours to trying. So I'm wondering if you might be able to run this on you workstation for me and paste the results?

#### Simulation Code

```
#!/usr/bin/env ruby

require 'date'
require 'json'
require 'tempfile'

$DEFAULT_RUNS = 1000


def tempfile(content)

  Tempfile.new("vagrant").tap do |temp|
    temp.binmode
    temp.write(content)
    temp.close
  end

end


def now
    DateTime.now.to_time.to_f
end


def simulate(runs=nil)
  runs = runs || ARGV[0] || $DEFAULT_RUNS
  runs = runs.to_i

  metrics = {
    :pass         => 0,
    :fail         => 0,
    :temp_files   => [],
    :iterations   => runs,
    :started_at   => now,
    :ended_at     => now,
    :run_duration => 0.0,
  }

  (1..runs).each do |index|

    begin
      tmp = tempfile("foo")

      metrics[:pass] += 1
      metrics[:temp_files] << {:filename => tmp.path, :exists? => File.exists?(tmp.path)}

    rescue
      metrics[:fail] += 1
    end
  end

  metrics[:ended_at] = now,
  metrics[:run_duration] = metrics[:ended_at] - metrics[:started_at]

  metrics
end

puts JSON.pretty_generate(simulate())
```

#### Simulation Usage

You'll need to find the path to the vagrant ruby executable to run the code above and then use something like this to execute the test.

Since I'm a linux guy here's how I would do it on linux. 

```
/opt/vagrant/embedded/bin/ruby test_tempfile.rb 10
```

This would run the simulation tests **10** times and report the temp file names and whether or not the temp file actually exists on disk when the function returns. From the docs I understand that the file will be deleted when the code goes out of scope, i.e. when either the garbage collected or when the program is terminated.

The output from the tests is a nice little JSON blob so that you could test for the file was actually deleted when the script terminates. I'll leave that for a task for you if you'd like to explore that. ;) I'll be happy to help if you'd like.

Here's what it looks like running from my MacOSX

```
{
  "pass": 10,
  "fail": 0,
  "temp_files": [
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1xk0tm7",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1e0nbdx",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-knef9a",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1fp5iy8",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-pvm5ad",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1nga0om",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1sfm2ah",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-1dfh1y2",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-12670yj",
      "exists?": true
    },
    {
      "filename": "/var/folders/36/_92163xs6tq2dmhr5y6613rm0000gn/T/vagrant20200605-54071-yrbwqc",
      "exists?": true
    }
  ],
  "iterations": 10,
  "started_at": 1591381460.020482,
  "ended_at": [
    1591381460.024221,
    0.0017809867858886719
  ],
  "run_duration": 0.0017809867858886719
}
```

```
/opt/vagrant/embedded/bin/ruby test_tempfile.rb 10
```

**Comment by chucknelson** [comment]
@codylane - Tried out this script with both versions of embedded ruby and can't get it to fail, regardless of how many iterations I specify :frowning: 

I'm able to reproduce the error pretty easily with `vagrant up` or `vagrant provision`, so at least it's consistently happening! I'll poke around more as I find time - if you have anything else you want me to test, let me know.

For reference, here are the ruby versions Vagrant is using:
```
# Vagrant 2.2.7
C:\HashiCorp\Vagrant\embedded\mingw64\bin\ruby --version
ruby 2.4.9p362 (2019-10-02 revision 67824) [x64-mingw32]

# Vagrant 2.2.8
C:\HashiCorp\Vagrant\embedded\mingw64\bin\ruby --version
ruby 2.6.6p146 (2020-03-31 revision 67876) [x64-mingw32]
```

I added a count of `exists?()` returning true or false, and it's always matching with the script, for example:
```
{
  ...
  "iterations": 50000,
  "started_at": 1591622628.2924821,
  "ended_at": [
    1591622885.432482,
    0.0
  ],
  "run_duration": 0.0,
  "pass": 50000,
  "exists": 50000,
  "fail": 0
}
```

**Comment by codylane** [comment]
Bummer, I was really hoping we see some failures. Thank you so much for testing this and providing the results.  :)

Ok, I've got another thing we can test for and I wanted to first try using the same code that vagrant-proxyconf uses which was the test result you pasted above.  That seems to be working nicely if you the running process is in the same scope.  

For the next test, we will actually simulate what `vagrant-proxyconf` does I'll create two separate `*.rb` files one that contains the `tempfile` code and one that overrides the code in a seperate file which will also be our test runner.  Hopefully we will see some failures from these new tests.  Let me see what I can whip up.

**Comment by codylane** [comment]
Actually, let's scratch the testing and go straight into full debugging. I have a feeling this maybe one that is hard to track down through testing. 

That said, are you familiar with debugging in ruby?  If not, maybe this can [help](https://learn.co/lessons/debugging-with-pry)?  You would have to install the `pry` gem into the vagrant ruby for this to work.  `gem.exe install pry` would do the trick as long as you can find the path to gem.exe command. 

Next we will add a few break points in the current version of the code and I assume you are using the latest version 2.0.7?  If so, then the next hurdle is finding out where the heck `vagrant-proxyconf` is installed on disk and then you'll add the following `require 'pry'; binding.pry` before the line [line](https://github.com/tmatilai/vagrant-proxyconf/blob/v2.0.7/lib/vagrant-proxyconf/action/configure_docker_proxy.rb#L144).  If all goes well, the next time you run `vagrant provision` assuming the VM is already running or if the VM fires up you will drop right into the debugger.  What you will be looking for is the current state of the variable `@docker_systemd_config` which you should just be able to copy and paste in the debugger and hit enter and it should tell you it's a *String* and a File that exists on disk.  You can test this by entering the following in the debugger.  If you wouldn't mind posting the results of this session perhaps we can find the root cause and get in a fix.

```
@docker_systemd_config.class.to_s
File.exists?(@docker_systemd_config)
```

**Comment by codylane** [comment]
@chucknelson - Okay, so I've got the code staged but I've got some failing tests and I'm going to step away from the computer for the evening.  I was really hoping to have this out the door today but I really don't want to cause any more problems for folks and I want to confirm that I'm not just getting to anxious to fix things. I'll be on this first thing tomorrow morning MDT.

**Comment by tmatilai** [comment]
> You bet my friend. I'm glad to help. I made a promise that I'm going to do my part volunteering where I can help and hopefully continue to maintain code that people use and like to use that I've contributed too. I don't even use this plugin anymore but I'm happy to continue to help maintain it.

Thank you so much for your work! You are amazing and very kind!

> As always, we cannot do it without without the assume open source community and people like you. We would love to continue having your help with this project and I'm sure others will appreciate it as well.

👍 

> I'd love to stay in touch and if you'd like to help join this project I'm more than happy to grant you some perms. We just need the blessing of the owner @tmatilai

I send invitation for @chucknelson to join the group of collaborators. Welcome onboard!

I myself have found very little time and energy for OSS work lately, and me neither am using Vagrant for this plugin, so I don't even follow the issues too closely. Especially the Windows and Docker bits have been contributed by other people.

So if I don't seem to respond to even pings here, you can always send me email directly. 😊

**Comment by codylane** [comment]
Released 2.0.8. closing request.  Thanks again for your help.
